### ISSUE TITLE
Generate for module

### ISSUE THREAD

**Issue by hayd** [MichaelHatherly/Lexicon.jl#126]
It'd be really great if there was a nice default api docs generating function for a module \*. Perhaps there is already and I just can't get it working?? ..The docs point towards [the build script in Sims](https://github.com/tshort/Sims.jl/blob/master/docs/build.jl)...but this fails for me with an exception\*\*...

Perhaps a `build(m::Module, dirs = [:src => :api])` (or similar) which looked for the .jl in the src directory and created an index for all the methods etc. defined there (under the api category). This seems to be what the Sims' build script does manually (is everyone supposed to reinvent this themselves?)...

\* By default doing this creates a markdown file with just the module title. No doubt I'm misreading the documentation, but I was hoping that

    Lexicon.save("foo.md", MyModule, Lexicon.Config())

would work. :s

I feel like I'm missing something fundamental here... :(

\*\* Can provide more info if it's a bug/can post as separate issue:
```
WARNING: both Interface and Base export "Docs"; uses of it in module Main must be qualified
ERROR: LoadError: UndefVarError: Docs not defined
 in find_loaded_packages at /Users/andy/.julia/v0.4/Docile/src/Cache/types.jl:44
...
```

**Comment by MichaelHatherly** [comment]
> It'd be really great if there was a nice default api docs generating function for a module

Agreed. I'll probably give the packages a bit of a facelift during 0.4/0.5 once the dust has settled and I know what can be ripped out and what should stay/get added. There was some work I did earlier in the year on https://github.com/MichaelHatherly/Lexicon.jl/tree/lexicon-next, but not really fit for consumption at the moment.

> is everyone supposed to reinvent this themselves

No, not really. I think some of that script may have been written prior to things being merged into Lexicon, but it's a while ago so I'm a bit hazy on the details.

> The docs point towards the build script in Sims...but this fails for me with an exception

Looks like that's because I've defined a type called ``Docs`` while ``Base`` has a module called ``Docs``... It'll need an explicit ``import Docile.Interface: Docs`` to sort it out I think.

> By default doing this creates a markdown file with just the module title.

Docile has probably not picked up any docstrings in ``MyModule``. What's the output of

```
julia> Docile.Cache.getraw(MyModule)
```

(Note that it must be a *real* package rather than just a source file so that Docile can automatically know that it's been imported. You'll have to use ``require("MyModule.jl")`` if it's not.)

**Comment by hayd** [comment]
Hmmm I get an exception running `Cache.getraw`, do you mind checking out https://github.com/JuliaGraphs/LightGraphs.jl/pull/109 ?

```
ERROR: MethodError: `exec` has no method matching exec(::Docile.Utilities.Head{:block}, ::Docile.Runner.State, ::Expr)
```

I think LightGraphs uses includes rather than requires... (is that the issue?)

Happy to help out with this where I can, being able to generate from just the module will be awesome :)

**Comment by MichaelHatherly** [comment]
> Hmmm I get an exception running Cache.getraw, do you mind checking out JuliaGraphs/LightGraphs.jl#109 ?

From a quick scan of the changes I can't see a particular signature that would be giving that error. It's due to Docile having a missing case for the partial-eval code in ``exec``. Best way to find it is to uncomment docstrings one at a time and see which causes it, after that it should be easy enough to resolve.

**Comment by hayd** [comment]
I'm now getting the following:
```
ERROR: ArgumentError: 'LightGraphs' is not a cached package.
```
Do I need to do something first?

**Comment by MichaelHatherly** [comment]
Odd... working on your PR branch, https://github.com/hayd/LightGraphs.jl/commit/96096706d49b90086784fb3096cdf64d1e8d8559, the following worked fine in 0.3:

```
julia> using LightGraphs

julia> Docile.Cache.getraw(LightGraphs)
Docile: updating package list...
Docile: caching 1 module from 'LightGraphs'.
ObjectIdDict with 143 entries:
...
```

This is on Julia ``0.3.12-pre+141`` rather than master. On master I can confirm the ``exec`` missing method error. So there's some parsing change I need to fix there.

> ```
> ERROR: ArgumentError: 'LightGraphs' is not a cached package.
> ```

What commands did you run to reproduce that?

**Comment by hayd** [comment]
@MichaelHatherly The same, but I can't seem to reproduce it anymore.

**Comment by MichaelHatherly** [comment]
Might have found it, Docile tries to automatically read external docstring sources by checking if a docstring is actually a file path. When it's a ``doc_str""`` it's failing to do it correctly. I'll have it fixed shortly.

**Comment by MichaelHatherly** [comment]
This diff seems to fix the ``exec`` problem:

**Comment by hayd** [comment]
:boom: yes it does!

**Comment by MichaelHatherly** [comment]
I'll have a tagged version ready shortly then.

**Comment by MichaelHatherly** [comment]
Tagged in https://github.com/JuliaLang/METADATA.jl/pull/3018.

**Comment by hayd** [comment]
Thanks!!! So now the api documentation is generated, whoop!

If I want these to be less flat I need to use require rather than includes? Would be great if these were grouped by file... ;)

**Comment by MichaelHatherly** [comment]
You could probably do that with ``filter``. Currently it just dumps everything in the module into one file, using ``require`` instead of ``include`` shouldn't have an effect on flatness.

The API for all this has been bugging me for a while &mdash; it's definitely not all that elegant. I did hack together something completely different in an unregistered package. A basic template-based system https://github.com/MichaelHatherly/Expresso.jl/blob/master/src/Templates.jl and an example build script https://github.com/MichaelHatherly/Expresso.jl/blob/master/doc/build.jl. I found it surprisingly nice to use.

**Comment by hayd** [comment]
Now this is fixed I think I can get the build.jl approach working... That looks very promising, but would be nice to remove the repeated bits (for each method) in the build.jl. I will have a go at this and ponder on what a good api might be. :)

**Comment by MichaelHatherly** [comment]
Yeah, the repeats were a bit unfortunate. Auto-generating better headers would definitely be nicer.

**Comment by hayd** [comment]
That said, sometimes it's nice to be explicit!

**Comment by MichaelHatherly** [comment]
Having some syntax to run an arbitrary formatting function over a docstring before splicing it into the output might work:

```jl
@file "foo.md",
"""
...
{{foobar |> formatter}}
...
"""
```

where formatter is some 1-arg user-defined function that returns the final string to splice in.

**Comment by hayd** [comment]
I think I would pull out the formatter, at least as high as:

```
@file "foo.md" formatter """
...
{{foobar}}

could also allow a list (of docstrings to be rendered):
{{foo, bar}}
"""
```

I can't actually seem to get this part working in 0.4, the `@doc` macro just doesn't return the docstring (it always returns nothing) Am I missing something here? Nothing like `@doc foo` or `@doc :(foo)` grabs the docstring... https://github.com/MichaelHatherly/Expresso.jl/blob/a500ffe1a977d89ffb26da2de885f77cef93f3e1/src/Templates.jl#L39.

---

Is there a neat way to pull out the signature info/Method and MethodTable (in a Markdown format), I'm guessing not (yet)... (I hate the inherent signature repetition forced into many docstrings currently.)

**Comment by MichaelHatherly** [comment]
> I think I would pull out the formatter, at least as high as:

Could do, perhaps a ``Dict`` rather than just a single symbol so other settings could be added.

> macro just doesn't return the docstring

That's quite weird... could you post a gist of what you've tried that hasn't been working?

> Is there a neat way to pull out the signature info/Method and MethodTable (in a Markdown format)

That would be cool to have. Since the ``Method`` objects are being stored in each ``FuncDoc`` and the ``Expr`` itself as well, we do have all the info needed. Just need to display it nicely. Getting well-formatted signatures can be a bit of a pain though given how complex some of them can become.

> I hate the inherent signature repetition forced into many docstrings currently.

Yeah, it does get old quite quickly.

**Comment by hayd** [comment]
> could you post a gist of what you've tried 

somehow restarting julia fixed it locally. However I still get nothing with the following gist https://gist.github.com/hayd/a9afa3e3049786d7c5a2 (Note: I tried tweaking it to allow passing lists inside `{{}}`)

**Comment by MichaelHatherly** [comment]
``p`` is a ``String`` here https://gist.github.com/hayd/a9afa3e3049786d7c5a2#file-build-jl-L26. So you'd need to do something like:

```jl
buildwriter(part, isdef) = isdef ?
    begin
        parts = Expr(:vect, [:(@doc($(parse(p)))) for p in split(part, r"\s*,\s*")]...)
        quote
            for p in $(esc(parts))
                writemime(file, "text/plain", p)
            end
        end
    end :
    :(print(file, $(esc(part))))
```

to get the expressions spliced into the correct places. This does mean you can't then have interpolation inside of ``{{...}}``, but that's not too limiting I think.

**Comment by hayd** [comment]
Ooooh, thanks! One possible way to do this then (updated with output): https://gist.github.com/hayd/a9afa3e3049786d7c5a2

**Comment by MichaelHatherly** [comment]
No prob.

Would be cool if the types in the signatures could be hyperlinks to their documentation. Not sure if there's a way to have links in code blocks though...

**Comment by hayd** [comment]
Well the good thing is, once they're generated it's trivial to set up. Please check out the LightTables PR for latest escapade... :)

(A few more tweaks, the main one ensure the methodtable only includes methods defined in the project.)

**Comment by hayd** [comment]
One annoying trip here is that defining functions also defined in base overwrites the old docstring. I suspect that's well known :)

```
help?> reverse
search: reverse reverse! reverseind SharedVector StridedVecOrMat StridedVector AbstractSparseVector

Base.reverse(v[, start=1[, stop=length(v)]])

   Return a copy of "v" reversed from start to stop.

julia> using LightGraphs

help?> LightGraphs.reverse
  (DiGraph only) Produces a graph where all edges are reversed from the original.

help?> reverse
search: reverse reverse! reverseind SharedVector StridedVecOrMat StridedVector AbstractSparseVector

  (DiGraph only) Produces a graph where all edges are reversed from the original.
```

**Comment by hayd** [comment]
check it: http://lightgraphsjl.readthedocs.org/en/latest/pathing/

**Comment by hayd** [comment]
One failure in this approach is if you have a previously used docstring in Base yours is appended... see https://lightgraphsjl.readthedocs.org/en/latest/operators/

I'm now a little confused, how does Lexicon fit in here? Is the idea that Lexicon can sit in the middle here and both genearte static/md sites and latex/pdf?

**Comment by hayd** [comment]
Thinking about it more, I think what may be preferable is (maybe as a macro)
```
md_generate([file1.md, file2.md, ...])
```
where each file1.md contains markdown + {{}} syntax. md_generates generates the expanded markdown into docs/_generated/file1.md (which is what's used on mkdocs; in which case makes sense to generate the mkdocs.yml). From the above we could then create the latex file (and the pdf).

Apologies for the brain dump!

**Comment by MichaelHatherly** [comment]
I think most of the problems could be alleviated if we had the ``@doc reverse(::DiGraph)``-type syntax for retrieving single ``Method`` docs. I've got an open PR, https://github.com/JuliaLang/julia/pull/11932, that adds most of the needed signature finding code. So I'll try adding some doc retrieval code on top of that.

**Comment by MichaelHatherly** [comment]
The system currently, [here](https://github.com/JuliaLang/julia/blob/b78659ea1c342346f600e685c87b78945f984024/base/docs/Docs.jl#L320), does allow using *values* to find single method docs, using ``@which``. It's a bit cumbersome when it's a complex type that's difficult to create though.

**Comment by MichaelHatherly** [comment]
OK, I think I've got something that will make it a bit easier to select just specific ``Method`` docs.

Could you try out https://github.com/JuliaLang/julia/pull/11932 with the following patch applied on it and see if it helps at all:

```diff
diff --git a/base/docs/Docs.jl b/base/docs/Docs.jl
index 90bcb2c..8b1968c 100644
--- a/base/docs/Docs.jl
+++ b/base/docs/Docs.jl
@@ -117,6 +117,15 @@ function doc(f::Function, m::Method)
     end
 end
 
+function methoddoc(ex)
+    f = esc(namify(ex.args[1]))
+    if any(x -> isexpr(x, :(::)), ex.args)
+        :(doc($f, which($f, $(esc(signature(ex))))))
+    else
+        :(doc($f, @which($(esc(ex)))))
+    end
+end
+
 catdoc() = nothing
 catdoc(xs...) = vcat(xs...)
 
@@ -317,7 +326,7 @@ end
 function docm(ex)
     isa(ex,Symbol) && haskey(keywords, ex) && return keywords[ex]
     isexpr(ex, :->) && return docm(ex.args...)
-    isexpr(ex, :call) && return :(doc($(esc(ex.args[1])), @which $(esc(ex))))
+    isexpr(ex, :call) && return methoddoc(ex)
     isexpr(ex, :macrocall) && (ex = namify(ex))
     :(doc($(esc(ex))))
 end
```

You'd the be able to, in your ``build.jl`` script, do something like

```jl

@file "foo.md" """
...
{{reverse(::DiGraph)}}
...
```

to get just the docs for ``reverse(::DiGraph)``, hopefully...

**Comment by MichaelHatherly** [comment]
Just a though regarding the ``,``-separated syntax you're using for splicing multiple docs at once. It'll clash with documenting specific methods with more than one argument since there'll be commas in the argument list as well as between each method, ie:

```jl
@file "foo.md" """
...
{{foo(::Int, ::Int), bar(::Float64, ::Int)}}
...
"""
```

How would separating each object by a newline instead? Something like:

```jl
@file "foo.md" """
...
{{
    foo(::Int, ::Int)
    bar(::Float64, ::Int)
}}
...
"""
```

I find that a little more readable and it avoids clashing uses of commas.

**Comment by MichaelHatherly** [comment]
Applying the newlines idea to that ``build.jl`` script in ``LightGraphs``, here https://gist.github.com/MichaelHatherly/0ea6962c01cd1858ea03, seems a lot clearer, if a little bit longer.

**Comment by MichaelHatherly** [comment]
> I'm now a little confused, how does Lexicon fit in here? Is the idea that Lexicon can sit in the middle here and both genearte static/md sites and latex/pdf?

At the moment Lexicon doesn't fit very well at all with 0.4's docsystem since there are some major internal differences between Docile and 0.4.

I wrote down some roadmap ideas for Docile/Lexicon [here](https://github.com/MichaelHatherly/Docile.jl/issues/147) that might make things a little more clear. I'll also move the template stuff into a Lexicon branch to make it a little more accessible.

**Comment by MichaelHatherly** [comment]
Created a new branch, https://github.com/MichaelHatherly/Lexicon.jl/tree/rewrite, with the templating code, but nothing else. Feel free to PR that automatic signature work if you'd like.

**Comment by hayd** [comment]
I think the newline idea is good.

The problem will doing a specific method is it won't show the other similar methods (that share the function with slightly different signature).... I'm not a fan of making new syntax for finding methods. This has come up in another issue here.

I think for the most part you want to paste in the whole function, which may have several different docstrings... perhaps you want something that for function `f` grabs:

    [docstring1 => [m1, m2], docstring2 => [m3], docstring2 => [m4]]

(where m1, m2, ... are f's methods\*) and prints

```
### function_name
``
m1 signature
m2
``
docstring1

``
m3
``
docstring2
...
```
Like I say, I'm not *really* sure how help on methods vs functions flies anymore...

I will checkout your PR. :)

\* these are only the methods defined in the package.

**Comment by MichaelHatherly** [comment]
Since the second argument to ``@file`` is just a plain string it mean you can interpolate and basically do anything you want. The result of the interpolation will then just be written to ``file`` the same as the string parts. So you could be something, just a quick sketch, like:

```jl
using Lexicon, LightGraphs

function submodules(mod::Module)
    out = Set([mod])
    for name in names(mod, true)
        if isdefined(mod, name)
            object = getfield(mod, name)
            validmodule(mod, object) && union!(out, submodules(object))
        end
    end
    out
end

validmodule(mod::Module, object::Module) = object ≠ mod && object ≠ Main
validmodule(::Module, other)             = false

function listmethods(func; from = Main, include_submodules = true)
    out = IOBuffer()
    for m in (include_submodules ? submodules(from) : Set([from]))
        if isdefined(m, :__META__)
            meta = m.__META__
            if haskey(meta, func)
                for each in meta[func].order
                    writemime(out, "text/plain", Docs.doc(func, each))
                end
            end
        end
    end
    takebuf_string(out)
end

@file "api.md",
"""

$(listmethods(
    reverse,
    from = LightGraphs,
    )
)

"""

```

Could actually just do away with ``{{...}}`` in that case...

**Comment by hayd** [comment]
How do I use `@file` for a list, say I want to do:

```
for fn in ["about.md"]
    @file joinpath("_generated", fn) open(readall, fn)
end
```

?

Isn't it problem with interpolation is conflicts with latex?

Aside: I haven't been able to get latex working, see https://github.com/JuliaGraphs/LightGraphs.jl/pull/114 (or css for that matter). :s

**Comment by MichaelHatherly** [comment]
Not sure if the looping would be possible, could be, but the actual value ``open(readall, fn)`` wouldn't be available until after the macro expander has run, unless you use ``eval``.

For latex input you could always just embed it in a section of the docs like so:

```jl
macro L_str(text) text end

@file "foo.md",
"""
...
$(L"...")
...
"""
```

**Comment by hayd** [comment]
I

**Comment by MichaelHatherly** [comment]
> I think the API shouldn't diverge too much from plain Markdown

It's just string interpolation, so really up to the user what they do with it. For most things ``{{}}`` will be the best way to go, in fact for getting docs for macros it's a lot easier since it won't have expanded before you what it to. What shouldn't end up happening is a whole new DSL inside ``{{}}``. It's good for simple things, but for anything more complex interpolation seems the sanest route to take.

> I don't understand why you can't do this with use regular functions

The macro is being used to get the current file with ``@__FILE__``, trying to use that from a function won't provide the correct location. Other than that it could probably be done fine using functions. Maybe there's a more elegant way to deal with paths?

**Comment by MichaelHatherly** [comment]
> Maybe there's a more elegant way to deal with paths?

Using ``do`` syntax does get most of the way there:

```jl
file(f) = f()

for f in ["a.md", "b.md"]
    file() do
        @show joinpath(Base.source_path(), f))
    end
end
```

So long as ``file`` is called at the toplevel then it does appear that the path is correct. That can then be used to do paths correctly.

This would require running ``eval`` to get docs out of ``{{}}`` though.

The equivalent of the macro syntax would then be something like:

```jl
using Lexicon

file("foo.md") do f
    """
    ...
    """
end
```

Personally I prefer the macro syntax more.

---

The looping problem could be solved by keeping each ``@file`` in a separate file:

``build.jl``
```jl
using Lexicon

for (name, file) in [("one.md", "a.jl"), ("two.md", "b.jl")]
    global path = joinpath(dirname(@__FILE__), "out", name)
    include(file)
end
```

That messy loop could be hidden in a macro.

``a.jl``
```jl
@file path,
"""
from a.jl
"""
```

``b.jl``
```jl
@file path,
"""
from b.jl
"""
```

**Comment by hayd** [comment]
I realise this is a compete hack, but is there a way to get the macro to work (you mentioned eval)

```
macro files(md_files...)
    for md in md_files
        @file(eval(:(joinpath("_generated", string($(md))))), eval(:(open(readall, $(md)))))
    end
end
```

**Comment by hayd** [comment]
The `__META__` trick fails with things like `readgml` which I docstringed as:
```
"docstring"
readgml
```

**Comment by MichaelHatherly** [comment]
> The ``__META__`` trick fails with things like ``readgml`` which I docstringed as

The line

```jl
for each in meta[func].order
```

is only iterating over the ``Method`` objects. When you do

```jl
"..."
foo
```

that associates ``"..."`` with the ``Function`` object whose docstring is stored in the ``.main`` field of the ``FuncDoc``.

You'd need to pull it out using something like this:

```jl
function listmethods(func; from = Main, include_submodules = true)
    out = IOBuffer()
    for m in (include_submodules ? submodules(from) : Set([from]))
        if isdefined(m, :__META__)
            meta = m.__META__
            if haskey(meta, func)
                writemime(out, "text/plain", meta[func].main) # <<<
                for each in meta[func].order
                    writemime(out, "text/plain", Docs.doc(func, each))
                end
            end
        end
    end
    takebuf_string(out)
end
```

**Comment by MichaelHatherly** [comment]
> I realise this is a compete hack, but is there a way to get the macro to work (you mentioned eval)

Yeah, not sure. I'll have a look at that this weekend if I get a chance.

**Comment by hayd** [comment]
Haven't a chance to test, but when I was thinking about this again I think you have to do something like:

    eval(:(@file $(joinpath("gen", file_name)) $(open(readall, file_name))
